compiler/rustc_infer/src/traits/project.rs

```diff
@@ -149,10 +169,41 @@ impl<'tcx> ProjectionCache<'_, 'tcx> {
             debug!("Not overwriting Recur");
             return;
         }
-        let fresh_key = map.insert(key, ProjectionCacheEntry::NormalizedTy(value));
+        let fresh_key =
+            map.insert(key, ProjectionCacheEntry::NormalizedTy { ty: value, complete: None });
         assert!(!fresh_key, "never started projecting `{:?}`", key);
     }
 
+    /// Mark the relevant projection cache key as having its derived obligations
+    /// complete, so they won't have to be re-computed (this is OK to do in a
+    /// snapshot - if the snapshot is rolled back, the obligations will be
+    /// marked as incomplete again).
+    pub fn complete(&mut self, key: ProjectionCacheKey<'tcx>, result: EvaluationResult) {
+        let mut map = self.map();
+        match map.get(&key) {
+            Some(&ProjectionCacheEntry::NormalizedTy { ref ty, complete: _ }) => {
+                info!("ProjectionCacheEntry::complete({:?}) - completing {:?}", key, ty);
+                let mut ty = ty.clone();
+                if result == EvaluationResult::EvaluatedToOk {
+                    ty.obligations = vec![];
```

Comment:
So, the docs above say "During evaluation, we will skip evaluating the cached sub-obligations in `ty` if this field is set." So, setting the obligations here to an empty vec is a bit redundant?

Seemingly, at the very least, the docs should be changed to reflect that the obligations are removed altogether. But maybe we can improve the type here to be encode the either-or state or "obligations or completion"

Reply:
I think an empty `Vec` is fine, and makes the type a little easier to work with. I'll update the docs to explain what's going on.